hardware/EcoDevices.cpp

```diff
@@ -122,365 +119,354 @@ bool CEcoDevices::WriteToHardware(const char *pdata, const unsigned char length)
-	// Check if a data source is connected to input 2
-	if (m_teleinfo2.PTEC !="----")
+	// Check EcoDevices firmware version and process pulse counters
+	sstr << "http://" << m_szIPAddress << ":" << m_usIPPort << "/status.xml";
+	if (m_status.hostname.empty()) m_status.hostname = m_szIPAddress;
+	if (HTTPClient::GET(sstr.str(), ExtraHeaders, sResult))
 	{
-		// PAPP only exist on some meter versions. If not present, we could approximate it as (current x 230V)
-		// but current is not available in JSON API, we need to switch to XML
-		if (root["T2_PAPP"].empty())
-			m_teleinfo2.PAPP = 0;
-		else
-			m_teleinfo2.PAPP = (unsigned long)(root["T2_PAPP"].asFloat());
-		if (m_teleinfo2.PTEC.substr(0,2) == "TH")
-		{
-			m_teleinfo2.rate="Tarif de base";
-		}
-		else if (m_teleinfo2.PTEC.substr(0,2) == "HC")
-		{
-			m_teleinfo2.rate="Heures Creuses";
-			m_pappHC=m_teleinfo2.PAPP;
-			m_pappHP=0;
-		}
-		else if (m_teleinfo2.PTEC.substr(0,2) == "HP")
-		{
-			m_teleinfo2.rate="Heures Pleines";
-			m_pappHC=0;
-			m_pappHP=m_teleinfo2.PAPP;
-		}
-		else if (m_teleinfo2.PTEC.substr(0,2) == "HN")
-		{
-			m_teleinfo2.rate="Heures Normales";
-			m_pappHC=m_teleinfo2.PAPP;
-			m_pappHP=0;
-		}
-		else if (m_teleinfo2.PTEC.substr(0,2) == "PM")
-		{
-			m_teleinfo2.rate="Pointe Mobile";
-			m_pappHC=0;
-			m_pappHP=m_teleinfo2.PAPP;
-		}
-		else
-		{
-			m_teleinfo2.rate="Unknown";
-			m_teleinfo2.tariff="Undefined";
-		}
+		_log.Log(LOG_NORM, "Ecodevices: Fetching counters and status data from %s", m_status.hostname.c_str());
+	}
+	else
+	{
+		_log.Log(LOG_ERROR, "EcoDevices: Error getting status.xml from EcoDevices%s!", m_status.hostname.c_str());
+		return;
+	}
 
-		// Basic subscription scheme
-		if (m_teleinfo2.PTEC == "TH..")
-		{
-			m_teleinfo2.tariff="Tarif de Base";
-			m_teleinfo2.BASE = (unsigned long)(root["T2_BASE"].asFloat());
-		}
+	// Process XML result
+	XMLdoc.Parse(sResult.c_str(), 0, TIXML_ENCODING_UTF8);
+	if (XMLdoc.Error())
+	{
+		_log.Log(LOG_ERROR, "Error parsing XML at /status.xml: %s", XMLdoc.ErrorDesc());
+		return;
+	}
 
-		// Peak/Off-Peak subscription scheme
-		else if ((m_teleinfo2.PTEC == "HC..") || (m_teleinfo2.PTEC == "HP.."))
+	// XML format changes dramatically between firmware versions. This code was developped for version 1.05.12
+        using namespace TinyXPath;
+	m_status.version = S_xpath_string(XMLdoc.RootElement(),"/response/version/text()").c_str();
+
+	#ifdef DEBUG_EcoDevices
+	std::string EncodedXML = sResult;
+        using boost::algorithm::replace_all;
+        replace_all(EncodedXML, "&",  "&amp;");
+        replace_all(EncodedXML, "\"", "&quot;");
+        replace_all(EncodedXML, "\'", "&apos;");
+        replace_all(EncodedXML, "<",  "&lt;");
+        replace_all(EncodedXML, ">",  "&gt;");
+        _log.Log(LOG_NORM, "DEBUG: XML output for /status.xml\n%s", EncodedXML.c_str());
+	#endif
+
+	major = boost::lexical_cast<int>(m_status.version.substr(0,m_status.version.find(".")).c_str());
```

Comment:
Are you 100% sure we always find a '.' ?
If yes, could you remove the boost:lexical_cast dependency (also the included header), and use
atoi(stdstring.c_str()) for int